File: adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovinInterstitialDelegate.m
// Copyright 2019 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "GADMAdapterAppLovinInterstitialDelegate.h"
#import "GADMAdapterAppLovinMediationManager.h"
#import "GADMAdapterAppLovinUtils.h"

@implementation GADMAdapterAppLovinInterstitialDelegate {
  /// AppLovin interstitial ad renderer to which the events are delegated.
  __weak GADMAdapterAppLovin *_parentRenderer;
}

#pragma mark - Initialization

- (nonnull instancetype)initWithParentRenderer:(nonnull GADMAdapterAppLovin *)parentRenderer {
  self = [super init];
  if (self) {
    _parentRenderer = parentRenderer;
  }
  return self;
}

#pragma mark - Ad Load Delegate

- (void)adService:(nonnull ALAdService *)adService didLoadAd:(nonnull ALAd *)ad {
  GADMAdapterAppLovin *parentRenderer = _parentRenderer;
  [GADMAdapterAppLovinUtils log:@"Interstitial did load ad: %@", ad];
  parentRenderer.interstitialAd = ad;
  [parentRenderer.connector adapterDidReceiveInterstitial:parentRenderer];
}

- (void)adService:(nonnull ALAdService *)adService didFailToLoadAdWithError:(int)code {
  GADMAdapterAppLovin *parentRenderer = _parentRenderer;
  [GADMAdapterAppLovinMediationManager.sharedInstance
      removeInterstitialZoneIdentifier:parentRenderer.zoneIdentifier];
  NSError *error = GADMAdapterAppLovinSDKErrorWithCode(code);
  [parentRenderer.connector adapter:parentRenderer didFailAd:error];
}

#pragma mark - Ad Display Delegate

- (void)ad:(nonnull ALAd *)ad wasDisplayedIn:(nonnull UIView *)view {
  [GADMAdapterAppLovinUtils log:@"Interstitial displayed"];
  GADMAdapterAppLovin *parentRenderer = _parentRenderer;
  [parentRenderer.connector adapterWillPresentInterstitial:parentRenderer];
}

- (void)ad:(nonnull ALAd *)ad wasHiddenIn:(nonnull UIView *)view {
  [GADMAdapterAppLovinUtils log:@"Interstitial hidden"];
  GADMAdapterAppLovin *parentRenderer = _parentRenderer;
  [GADMAdapterAppLovinMediationManager.sharedInstance
      removeInterstitialZoneIdentifier:parentRenderer.zoneIdentifier];
  id<GADMAdNetworkConnector> strongConnector = parentRenderer.connector;
  [strongConnector adapterWillDismissInterstitial:parentRenderer];
  [strongConnector adapterDidDismissInterstitial:parentRenderer];
}

- (void)ad:(nonnull ALAd *)ad wasClickedIn:(nonnull UIView *)view {
  [GADMAdapterAppLovinUtils log:@"Interstitial clicked"];
  GADMAdapterAppLovin *parentRenderer = _parentRenderer;
  id<GADMAdNetworkConnector> strongConnector = parentRenderer.connector;
  [GADMAdapterAppLovinUtils log:@"Interstitial clicked"];
  [strongConnector adapterDidGetAdClick:parentRenderer];
  [strongConnector adapterWillLeaveApplication:parentRenderer];
}

#pragma mark - Video Playback Delegate

- (void)videoPlaybackBeganInAd:(nonnull ALAd *)ad {
  [GADMAdapterAppLovinUtils log:@"Interstitial video playback began"];
}

- (void)videoPlaybackEndedInAd:(nonnull ALAd *)ad
             atPlaybackPercent:(nonnull NSNumber *)percentPlayed
                  fullyWatched:(BOOL)wasFullyWatched {
  [GADMAdapterAppLovinUtils log:@"Interstitial video playback ended at playback percent: %lu%%",
                                (unsigned long)percentPlayed.unsignedIntegerValue];
}

@end

File: adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovinInterstitialDelegate.swift
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import AppLovinSDK
import Foundation
import GoogleMobileAds

@MainActor
@objc(GADMAdapterAppLovinInterstitialDelegate)
public class GADMAdapterAppLovinInterstitialDelegate: NSObject, @preconcurrency ALAdLoadDelegate,
  @preconcurrency ALAdDisplayDelegate, @preconcurrency ALAdVideoPlaybackDelegate
{

  private weak var parentRenderer: GADMAdapterAppLovin?

  @objc public init(parentRenderer: GADMAdapterAppLovin) {
    self.parentRenderer = parentRenderer
    super.init()
  }

  // MARK: - Ad Load Delegate
  public func adService(_ adService: ALAdService, didLoad ad: ALAd) {
    guard let parentRenderer else { return }
    let isMultipleAdsEnabled = GADMAdapterAppLovinUtils.isMultipleAdsLoadingEnabled()
    if isMultipleAdsEnabled {
      GADMAdapterAppLovinMediationManager.sharedInstance.removeInterstitialZoneIdentifier(
        parentRenderer.zoneIdentifier ?? "")
    }
    GADMAdapterAppLovinUtils.log("Interstitial did load ad: \(ad)")
    parentRenderer.interstitialAd = ad
    parentRenderer.connector?.adapterDidReceiveInterstitial(parentRenderer)
  }

  public func adService(_ adService: ALAdService, didFailToLoadAdWithError code: Int32) {
    guard let parentRenderer else { return }
    GADMAdapterAppLovinMediationManager.sharedInstance.removeInterstitialZoneIdentifier(
      parentRenderer.zoneIdentifier ?? "")
    let error = GADMAdapterAppLovinUtils.sdkError(withCode: Int(code))
    parentRenderer.connector?.adapter(parentRenderer, didFailAd: error)
  }

  // MARK: - Ad Display Delegate
  public func ad(_ ad: ALAd, wasDisplayedIn view: UIView) {
    GADMAdapterAppLovinUtils.log("Interstitial displayed")
    guard let parentRenderer else { return }
    parentRenderer.connector?.adapterWillPresentInterstitial(parentRenderer)
  }

  public func ad(_ ad: ALAd, wasHiddenIn view: UIView) {
    GADMAdapterAppLovinUtils.log("Interstitial hidden")
    guard let parentRenderer else { return }
    GADMAdapterAppLovinMediationManager.sharedInstance.removeInterstitialZoneIdentifier(
      parentRenderer.zoneIdentifier ?? "")
    let connector = parentRenderer.connector
    connector?.adapterWillDismissInterstitial(parentRenderer)
    connector?.adapterDidDismissInterstitial(parentRenderer)
  }

  public func ad(_ ad: ALAd, wasClickedIn view: UIView) {
    GADMAdapterAppLovinUtils.log("Interstitial clicked")
    guard let parentRenderer else { return }
    let connector = parentRenderer.connector
    connector?.adapterDidGetAdClick(parentRenderer)
    connector?.adapterWillLeaveApplication(parentRenderer)
  }

  // MARK: - Video Playback Delegate
  public func videoPlaybackBegan(in ad: ALAd) {
    GADMAdapterAppLovinUtils.log("Interstitial video playback began")
  }

  public func videoPlaybackEnded(
    in ad: ALAd, atPlaybackPercent percentPlayed: NSNumber, fullyWatched wasFullyWatched: Bool
  ) {
    GADMAdapterAppLovinUtils.log(
      "Interstitial video playback ended at playback percent: \(percentPlayed.uintValue)%")
  }
}
